### PR TITLE
Navigation model refactor to show subnav hierarchy

### PR DESCRIPTION
**Pull request by NataliaLKB**
## What does this change?
This PR is motivated by two things. 

1. We want to visually distinguish between secondary and tertiary items. Right now, that isn't possible.
2. Navigation should be organised more as a tree, but that isn't currently happening. This refactor makes the nav much more of a tree structure. 

Thank you @TBonnin for all your help!!

## Screenshots
**before**:
image

**after**:
image

## What is the value of this and can you measure success?
The currently behaviour and visuals shouldn't change much, but hopefully the code is easier to follow.

## Does this affect other platforms - Amp, Apps, etc?
Nope

## Does this affect GLabs Paid Content Pages? Should it have support for Paid Content?

Nope

## Tested in CODE?
- [x] :crossed_fingers:

### CODE REVIEW

File: common/app/views/fragments/nav/subNav.scala.html [hunk 1]
```diff
@@ -1,16 +1,18 @@
...
 
+@(navMenu: NavMenu, isFooter: Boolean = false)(implicit request: RequestHeader)
+
 
-@subNavItem(link: NavLink, id: String) = {
+@subNavItem(link: NavLink, url: String, isParent: Boolean = false, isChild: Boolean = false) = {
     <li class="@RenderClasses(Map(
-            "subnav__item--current-section" -> (id == link.uniqueSection)
+            "subnav__item--current-section" -> (url == link.url),
+            "subnav__item--parent" -> isParent,
```
Comment (NataliaLKB) [inline common/app/views/fragments/nav/subNav.scala.html]: this class and the class below are in anticipation for further styling changes

File: common/app/navigation/Navigation.scala [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+import _root_.model.Page
+import NavLinks._
+
+import scala.annotation.tailrec
+
+sealed trait NavNode {
+  def children: Seq[NavLink]
+}
+
+case class NavLink(
```
Comment (regiskuckaertz) [inline common/app/navigation/Navigation.scala]: Should this extend `NavNode`?

File: common/app/navigation/Navigation.scala [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        case head :: tail if (head.url == url) => Some(head)
+        case head :: tail => find(tail ++ head.children)
+      }
+    }
+    find(children)
+  }
+
+  def findParentByUrl(maybeCurrentNavLink: Option[NavLink]): Option[NavLink] = {
+
+    maybeCurrentNavLink.map(currentNavLink => {
```
Comment (regiskuckaertz) [inline common/app/navigation/Navigation.scala]: I suggest you leave that out for better composability; this method should not have to deal with all the ceremony of unfolding an option and providing an alternative. It is just a changing a `findParentByUrl(maybeLink)` to `maybeLink.flatMap(findParentByUrl)`.

File: common/app/navigation/Navigation.scala [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+  def subNavSections: Subnav = root.getSubnav(currentNavLink, currentParent, currentPillar)
+}
+
+object NavMenu {
+
+  def apply(page: Page, edition: Edition): NavMenu = NavMenu(page, NavRoot(edition), edition)
+
+  def apply(edition: Edition): SimpleMenu = SimpleMenu(NavRoot(edition))
+
+  def getSectionOrPageUrl(page: Page, edition: Edition): String = {
```
Comment (NataliaLKB) [inline common/app/navigation/Navigation.scala]: This stuff is a bit ugly, but its just porting over the same logic as before and I would rather not rewrite this too right now

**Comment by NataliaLKB** [comment]
@TBonnin @regiskuckaertz I am going to test this in CODE this afternoon, but I think its just about good to go! If you could have another look at the code it would be much appreciated.

File: common/app/navigation/Navigation.scala [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    @tailrec
+    def find(children: Seq[NavLink]): Option[NavLink] = {
+      children match {
+        case Nil => None
+        case head :: tail if (head.url == url) => Some(head)
+        case head :: tail => find(tail ++ head.children)
+      }
+    }
+
+    val maybeNavLink = find(children)
```
Comment (regiskuckaertz) [inline common/app/navigation/Navigation.scala]: `find(children).orElse(find(getOtherPillarsFromEdition(edition))` 🤗

**Review by regiskuckaertz — COMMENTED** [review]
Sorry for my comments, they were mostly just styling 😓 Awesome work! I'll let Thomas give you the final go.

File: common/app/navigation/Navigation.scala [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+  def findParentByCurrentNavLink(currentNavLink: NavLink, edition: Edition): Option[NavLink] = {
+    // Football is currently in the News Pillar and the Sport pillar, however we don't want the parent to be News.
+    def footballDoesNotBelongInNews(parentTitle: String): Boolean = {
+      currentNavLink.title == "football" && parentTitle == "news"
+    }
+
+    @tailrec
+    def find(children: Seq[NavLink]): Option[NavLink] = {
+      children match {
+        case Nil => find(getOtherPillarsFromEdition(edition))
```
Comment (TBonnin) [inline common/app/navigation/Navigation.scala]: Same question here. Why do we need to lookup in other edition?

File: common/app/navigation/Navigation.scala [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        case Nil => None
+        case head :: tail if (head.url == url) => Some(head)
+        case head :: tail => find(tail ++ head.children)
+      }
+    }
+    // If the link isn't found within the current edition, check other editions
+    find(children).orElse(find(getOtherPillarsFromEdition(edition)))
+  }
+
+  def findParentByCurrentNavLink(currentNavLink: NavLink, edition: Edition): Option[NavLink] = {
```
Comment (TBonnin) [inline common/app/navigation/Navigation.scala]: `parent`?

File: common/app/navigation/Navigation.scala [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    @tailrec
+    def find(children: Seq[NavLink]): Option[NavLink] = {
+      children match {
+        case Nil => find(getOtherPillarsFromEdition(edition))
+        case head :: tail if (head == currentNavLink) => Some(head) // If primary link, use that
+        case head :: tail if (footballDoesNotBelongInNews(head.title)) => find(tail ++ head.children)
+        case head :: tail if head.children.contains(currentNavLink) => Some(head)
+        case head :: tail => find(tail ++ head.children)
```
Comment (TBonnin) [inline common/app/navigation/Navigation.scala]: Can this be simplified:
```
...
case head :: tail if (head == currentNavLink || head.children.contains(currentNavLink) && !isFootballInNews(head)) => Some(head) 
case head :: tail => find(tail ++ head.children)
```

File: common/app/navigation/Navigation.scala [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+      }
+    }
+    // If the parent isn't found within the current edition, check other editions
+    find(children).orElse(find(getOtherPillarsFromEdition(edition)))
+  }
+
+  def getPillar(pillars: Seq[NavLink], currentParent: NavLink, edition: Edition): NavLink = {
```
Comment (TBonnin) [inline common/app/navigation/Navigation.scala]: Do we need to pass the `pillars` here? Isn't `NavRoot.children` the pillars?

File: common/app/navigation/Navigation.scala [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+    if(pillars.contains(currentParent)) {
+      currentParent
+    } else {
+      // TODO: should we have default pillar that is not UK (like a blank one)?
+      findParentByCurrentNavLink(currentParent, edition).getOrElse(ukNewsPillar)
+    }
+  }
+
+  def getSubnav(currentNavLink: Option[NavLink], currentParent: NavLink, currentPillar: NavLink): Subnav = {
```
Comment (TBonnin) [inline common/app/navigation/Navigation.scala]: `currentParent` and `currentPillar` can be deducted based on the current `NavLink`. Why do we have to pass them as param?

File: common/app/navigation/Navigation.scala [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+    val currentNavHasChildren = currentNavLink.map(navLink => navLink.children.nonEmpty).getOrElse(false)
+    val parentIsPillar =  currentNavLink.contains(currentPillar)
+
+    val showParent = if (parentIsPillar) false else true
```
Comment (TBonnin) [inline common/app/navigation/Navigation.scala]: `val showParent = !parentIsPillar` ?

File: common/app/navigation/Navigation.scala [hunk 12]
```diff
@@ -0,0 +1,157 @@
+package navigation
+
+import common.{Edition, Navigation, editions}
+import _root_.model.Page
+import NavLinks._
+
+import scala.annotation.tailrec
+
+case class Subnav(parent: Option[NavLink], children: Option[Seq[NavLink]], hasSecondaryParent: Boolean = false, hasTertiary: Boolean = false)
```
Comment (TBonnin) [inline common/app/navigation/Navigation.scala]: I am not sure I understand the purpose of `hasSecondaryParent` and `hasTertiary`? Can you clarify?

**Review by TBonnin — CHANGES_REQUESTED** [review]
I left quite a lot of feedback. let me know if you want to go through them in person. It might be easier

File: common/app/navigation/Navigation.scala [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+
+sealed trait NavNode {
+  def children: Seq[NavLink]
+}
+
+case class NavLink(
+  title: String,
+  url: String,
+  longTitle: String = "",
+  // TODO: Shouldn't need iconName. Remove, and just make the first NavLink on mobile
```
[the remark this answers is not in the document]
Comment (NataliaLKB) [reply common/app/navigation/Navigation.scala]: Yep! But I don't want to do it within this PR

File: common/app/navigation/Navigation.scala [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+    @tailrec
+    def find(children: Seq[NavLink]): Option[NavLink] = {
+      children match {
+        case Nil => None
+        case head :: tail if (head.url == url) => Some(head)
+        case head :: tail => find(tail ++ head.children)
+      }
+    }
+    // If the link isn't found within the current edition, check other editions
+    find(children).orElse(find(getOtherPillarsFromEdition(edition)))
```
[the remark this answers is not in the document]
Comment (NataliaLKB) [reply common/app/navigation/Navigation.scala]: We need to look into other editions for the case when you end up in a section like `cricket` (say from Google) and you are in the US. It is not within the US sports pillar, but it should still show within a sports pillar

File: common/app/navigation/Navigation.scala (hunk 9 above)
Comment (NataliaLKB) [reply common/app/navigation/Navigation.scala]: Good point!

File: common/app/navigation/Navigation.scala [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+  longTitle: String = "",
+  // TODO: Shouldn't need iconName. Remove, and just make the first NavLink on mobile
+  iconName: String = "",
+  children: Seq[NavLink] = Nil
+) extends NavNode
+
+
+case class NavRoot private(children: Seq[NavLink], otherLinks: Seq[NavLink], brandExtensions: Seq[NavLink]) extends NavNode {
+
+  def getOtherPillarsFromEdition(edition: Edition): Seq[NavLink] = {
```
Comment (TBonnin) [inline common/app/navigation/Navigation.scala]: naming?

File: common/app/navigation/Navigation.scala [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+  }
+
+  def getPillar(currentParent: Option[NavLink], edition: Edition): Option[NavLink] = {
+    if(otherLinks.contains(currentParent.getOrElse(None))) {
+      None
+    } else if(children.contains(currentParent.getOrElse(None))) {
+      currentParent
+    } else {
+      currentParent.map( link => findParent(link, edition)).getOrElse(Some(ukNewsPillar))
+    }
```
Comment (TBonnin) [inline common/app/navigation/Navigation.scala]: `if` inside of `map`

File: common/app/navigation/Navigation.scala [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+    val parentIsPillar =  currentParent.equals(currentPillar)
+    def currentNavIsPillar = currentNavLink.equals(currentPillar)
+
+    val parent = if(currentNavHasChildren & !currentNavIsPillar) currentNavLink else if (parentIsPillar) None else currentParent
+
+    val childrenToShow = if (currentNavHasChildren) currentNavLink else currentParent
+    val children = childrenToShow.map( navLink => navLink.children ).getOrElse(Nil)
+
+    Subnav(parent, children, parent.isDefined || children.nonEmpty, parent.isDefined)
+  }
```
Comment (TBonnin) [inline common/app/navigation/Navigation.scala]: move it to `NavMenu` and make it private

File: common/app/navigation/Navigation.scala [hunk 18]
```diff
@@ -0,0 +1,160 @@
+package navigation
+
+import common.{Edition, Navigation, editions}
+import _root_.model.Page
+import NavLinks._
+
+import scala.annotation.tailrec
+
+case class Subnav(parent: Option[NavLink], children: Seq[NavLink], showSubnav: Boolean, hasParent: Boolean = false)
```
Comment (TBonnin) [inline common/app/navigation/Navigation.scala]: hasParent == parent.None

**Comment by NataliaLKB** [comment]
I have checked this out on CODE, and it looks good!

File: common/app/navigation/Navigation.scala [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+
+
+case class NavRoot private(children: Seq[NavLink], otherLinks: Seq[NavLink], brandExtensions: Seq[NavLink]) extends NavNode {
+
+  def getChildrenFromOtherEditions(edition: Edition): Seq[NavLink] = {
+    Edition.others(edition).flatMap( edition => {
+      val root = NavRoot(edition)
+
+      root.children ++ root.otherLinks
+    })
```
Comment (TBonnin) [inline common/app/navigation/Navigation.scala]: neatpicking: unecessary `{}`

File: common/app/navigation/Navigation.scala [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+  }
+
+  def getPillar(currentParent: Option[NavLink], edition: Edition): Option[NavLink] = {
+    currentParent.flatMap( parent => {
+      if(otherLinks.contains(parent)) {
+        None
+      } else if (children.contains(parent)) {
+        currentParent
+      } else findParent(parent, edition).orElse(Some(ukNewsPillar))
+    })
```
Comment (TBonnin) [inline common/app/navigation/Navigation.scala]: same. You don't need the inner `{}` in the `flatMap`

File: common/app/navigation/Navigation.scala [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+  def currentPillar: Option[NavLink] = root.getPillar(currentParent, edition)
+  def subNavSections: Subnav = NavMenu.getSubnav(currentNavLink, currentParent, currentPillar)
+}
+
+object NavMenu {
+
+  def apply(page: Page, edition: Edition): NavMenu = NavMenu(page, NavRoot(edition), edition)
+  def apply(edition: Edition): SimpleMenu = SimpleMenu(NavRoot(edition))
+
+  def getSectionOrPageUrl(page: Page, edition: Edition): String = {
```
Comment (TBonnin) [inline common/app/navigation/Navigation.scala]: you can make this function as well as `getSubnav` below package private (`private[navigation] def ...`)

**Review by TBonnin — APPROVED** [review]
+💯for the tests. Thank you Natalia

**Pull request merged**